srtcore/sync.cpp

```diff
@@ -0,0 +1,168 @@
+    if (QueryPerformanceFrequency(&ccf))
+        frequency = ccf.QuadPart / 1000000; // counts per microsecond
+
+#elif defined(TIMING_USE_MACH_ABS_TIME)
+
+    mach_timebase_info_data_t info;
+    mach_timebase_info(&info);
+    frequency = info.denom * int64_t(1000) / info.numer;
+
+#elif defined(IA32) || defined(IA64) || defined(AMD64)
+    uint64_t t1, t2;
+
+    rdtsc(t1);
+    timespec ts;
+    ts.tv_sec  = 0;
+    ts.tv_nsec = 100000000;
+    nanosleep(&ts, NULL);
+    rdtsc(t2);
+
+    // CPU clocks per microsecond
+    frequency = int64_t(t2 - t1) / 100000;
+#endif
+
+    return frequency;
+}
+
+static const int64_t s_cpu_frequency = get_cpu_frequency();
+
+} // namespace sync
+} // namespace srt
+
+template <>
+uint64_t srt::sync::TimePoint<srt::sync::steady_clock>::us_since_epoch() const
+{
+    return m_timestamp / s_cpu_frequency;
+}
+
+template <>
+srt::sync::Duration<srt::sync::steady_clock> srt::sync::TimePoint<srt::sync::steady_clock>::time_since_epoch() const
+{
+    return srt::sync::Duration<srt::sync::steady_clock>(m_timestamp);
+}
+
+srt::sync::TimePoint<srt::sync::steady_clock> srt::sync::steady_clock::now()
+{
+    uint64_t x = 0;
+    rdtsc(x);
+    return TimePoint<steady_clock>(x);
+}
+
+long long srt::sync::count_microseconds(const steady_clock::duration& t) { return t.count() / s_cpu_frequency; }
+
+long long srt::sync::count_milliseconds(const steady_clock::duration& t) { return t.count() / s_cpu_frequency / 1000; }
+
+long long srt::sync::count_seconds(const steady_clock::duration& t) { return t.count() / s_cpu_frequency / 1000000; }
+
+srt::sync::steady_clock::duration srt::sync::microseconds_from(long t_us)
+{
+    return steady_clock::duration(t_us * s_cpu_frequency);
+}
+
+srt::sync::steady_clock::duration srt::sync::milliseconds_from(long t_ms)
+{
+    return steady_clock::duration((1000 * t_ms) * s_cpu_frequency);
+}
+
+srt::sync::steady_clock::duration srt::sync::seconds_from(long t_s)
+{
+    return steady_clock::duration((1000000 * t_s) * s_cpu_frequency);
+}
+
+std::string srt::sync::FormatTime(const steady_clock::time_point& timestamp)
+{
+    const time_t                   now_s         = ::time(NULL); // get current time in seconds
+    const steady_clock::time_point now_timestamp = steady_clock::now();
+    const int64_t                  delta_us      = count_microseconds(timestamp - now_timestamp);
+    const int64_t                  delta_s =
+        floor((static_cast<int64_t>(now_timestamp.us_since_epoch() % 1000000) + delta_us) / 1000000.0);
+    const time_t tt = now_s + delta_s;
+    struct tm    tm = SysLocalTime(tt); // in seconds
+    char         tmp_buf[512];
+    strftime(tmp_buf, 512, "%X.", &tm);
+
+    ostringstream out;
+    out << tmp_buf << setfill('0') << setw(6) << (timestamp.us_since_epoch() % 1000000);
+    return out.str();
+}
+
+int srt::sync::SyncEvent::wait_for(pthread_cond_t* cond, pthread_mutex_t* mutex, const Duration<steady_clock>& rel_time)
```

Comment:
The implementation requires that when pthread_cond_t is created, it respects requirements imposed by CLOCK_MONOTONIC usage but currently, there are no guarantees that that is done or hints that it should be done. I don't know if we want to start wrapping things into objects just for this, but maybe 'sync::SyncEvent()' could have another static helper method that would properly create and return pthread_cond_t instance just to encourage good usage?

Reply:
That is a good point.
The next PR #800 replaces all "hanging" congitional variables with the self-contained `SyncEvent` object. So you just create one `SyncEvent` instead of CV + mutex pair, add call:
```
int srt::sync::SyncEvent::wait_for(const Duration<steady_clock>& rel_time);
```

Therefore I suggest to keep it this way for now. This should also produce a cleaner diff for the next PR.

Reply:
I highly recommend then that you confront it with #1004, as there's also a wrapper for CV provided there. We need to agree upon one version, and for starters I'd at least recommend that we agree upon the used API, otherwise we'll have an irresolvable mess later.

Reply:
Eventually, 
`int SyncEvent::wait_for(pthread_cond_t* cond, pthread_mutex_t* mutex, const Duration<steady_clock>& rel_time)`
will become
`int SyncEvent::wait_for(const Duration<steady_clock>& rel_time)`

This will happen when SyncEvent will own the CV and mutex. 
Then this will correspond to
`bool CCondDelegate::wait_for(uint64_t delay)`